### PR TITLE
Support Webpack 4

### PR DESCRIPTION
**Pull request by jkrems**
This is definitely a breaking change but I hope it'll be useful for yourself or other people who are trying to use this plugin with webpack 4.

* plugin: Upgrade to webpack 4
* plugin: Fix harmony import detection
* lib: Resolve deprecation warnings

---
_This PR was started by: [git wf pr](https://github.groupondev.com/InteractionTier/workflow-cli/releases/tag/v1.7.0)_

### CODE REVIEW

**Comment by jkrems** [comment]
I also tested this branch against our internal asset pipeline and the old webpack 3 test case now passes with webpack 4.

**Comment by indutny** [comment]
Awesome patch!

@jkrems would you be interested in becoming a maintainer of this project?

File: lib/shake/plugin.js [hunk 1]
```diff
@@ -149,7 +156,7 @@ ShakePlugin.prototype.mapModule = function mapModule(state, compilation,
   }
 
   const isImported = module.reasons.some((reason) => {
-    return reason.dependency.type === 'harmony import';
+    return /^harmony import/.test(reason.dependency.type);
```
Comment (kurtextrem) [inline lib/shake/plugin.js]: ```reason.dependency.type.indexOf('harmony import') === 0```
or more recent syntax:
```reason.dependency.type.startsWith('harmony import')```

Imo better than using a RegExp, but up to you.

Comment (jkrems) [reply lib/shake/plugin.js]: Good point. Especially "startsWith" is a lot clearer. Always forget about that one when interacting with modern code. :D

File: test/plugin-test.js [hunk 2]
```diff
@@ -30,6 +30,8 @@ describe('webpack-common-shake', () => {
       bail: true,
       entry: path.join(FIXTURES_DIR, file),
       output: {
+        library: '$shakeResult',
```
Comment (jkrems) [inline test/plugin-test.js]: This hack works around a change in the way webpack wraps the modules in its runtime. The expression now evaluates to "true" by default, no longer to the exports of the entry point. By using this hack we get the side effect of the assignment to `window.$shakeResult` which gives us the previous value.

**Comment by jkrems** [comment]
@indutny Sure! I'm not sure how much time I'd be able to spend on it on a regular basis but we'll be actively using it and I'm happy to help react to the issue tracker / address bugs and regressions / etc..

**Comment by indutny** [comment]
@jkrems thank you so much! I've added you to github collaborators. What is your npm id?

**Comment by jkrems** [comment]
Got the invite! My id on npm is `jkrems`.

**Comment by indutny** [comment]
Yay! Added you to npm owners too. Please consider enabling 2FA on npm, if you haven't already!

Thanks again!

**Comment by indutny** [comment]
@jkrems there're not that many collaborators at the moment, so green CI might be considered a sign-off. Otherwise, it is just as you said: squash, `npm version xxx` and push + publish.

**Comment by jkrems** [comment]
Got it. Will try to publish this change as a major bump.

**Comment by jkrems** [comment]
Published as v2.0.0

**Pull request merged**